### PR TITLE
Use thread pool for assign_regions

### PR DESCRIPTION
**Pull request by alannotnerd**
Thread pool is required when the api caller given a large amount of subregions, which crossbeam doesn't and won't support.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 .vscode
 **/*.html
 .DS_Store
+.nlsp-settings/
```
Comment (kunxian-xia) [inline .gitignore]: Can you revert this?

File: halo2_proofs/tests/plonk_api.rs [hunk 2]
```diff
@@ -1110,3 +1110,620 @@ fn plonk_api() {
...
+        fn raw_multiply<F>(
+            &self,
+            layouter: &mut impl Layouter<FF>,
+            f: F,
+        ) -> Result<(Cell, Cell, Cell), Error>
+        where
+            F: FnMut() -> Value<(Assigned<FF>, Assigned<FF>, Assigned<FF>)>;
+        fn raw_add<F>(
+            &self,
+            layouter: &mut impl Layouter<FF>,
```
Comment (kunxian-xia) [inline halo2_proofs/tests/plonk_api.rs]: A better solution would be reuse these structs defined in the plonk_api test. You can move these structs from the plonk_api test scope out to the entire module level.

File: halo2_proofs/tests/plonk_api.rs [hunk 3]
```diff
@@ -1110,3 +1110,620 @@ fn plonk_api() {
...
+             *   A         B      ...  sl
+             * [
+             *   instance  0      ...  0
+             *   a         a      ...  0
+             *   a         a^2    ...  0
+             *   a         a      ...  0
+             *   a         a^2    ...  0
+             *   ...       ...    ...  ...
+             *   ...       ...    ...  instance
+             *   ...       ...    ...  a
```
Comment (kunxian-xia) [inline halo2_proofs/tests/plonk_api.rs]: We could move these ops into another function `PlonkConfig::construct(meta)` and use it both in the `plonk_api` test and the new test.

**Comment by alannotnerd** [comment]
@kunxian-xia Your comments are resolved. May you want to take a look again?

**Comment by kunxian-xia** [comment]
Can you fix the ci errors and resolve the conflicts?

File: halo2_proofs/src/circuit/floor_planner/single_pass.rs [hunk 4]
```diff
@@ -215,6 +215,7 @@ impl<'a, F: Field, CS: Assignment<F> + 'a> Layouter<F> for SingleChipLayouter<'a
         N: Fn() -> NR,
         NR: Into<String>,
     {
+        use rayon::prelude::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator};
```
Comment (kunxian-xia) [inline halo2_proofs/src/circuit/floor_planner/single_pass.rs]: It's better to move this import directive to head of the file.

**Review by kunxian-xia — APPROVED** [review]

**Pull request merged**